consensus/consensus.go

```diff
@@ -585,11 +622,68 @@ func (consensus *Consensus) Prepare(chain consensus_engine.ChainReader, header *
 	return nil
 }
 
-// AccumulateRewards credits the coinbase of the given block with the mining
+// accumulateRewards credits the coinbase of the given block with the mining
 // reward. The total reward consists of the static block reward and rewards for
 // included uncles. The coinbase of each uncle block is also rewarded.
-func accumulateRewards(config *params.ChainConfig, state *state.DB, header *types.Header) {
-	// TODO: implement mining rewards
+func (consensus *Consensus) accumulateRewards(
+	config *params.ChainConfig, state *state.DB, header *types.Header,
+) {
+	if header.ParentHash == (common.Hash{}) {
+		// This block is a genesis block,
+		// without a parent block whose signer to reward.
+		return
+	}
+	if consensus.ChainReader == nil {
+		utils.GetLogInstance().Error("accumulateRewards: ChainReader is nil")
+		return
+	}
+	parent := consensus.ChainReader.GetBlockByHash(header.ParentHash)
+	if parent == nil {
+		utils.GetLogInstance().Error(
+			"accumulateRewards: cannot retrieve parent block",
+			"curHash", header.Hash().Hex(),
+			"parentHash", header.ParentHash.Hex(),
+		)
+		return
+	}
+	mask, err := bls_cosi.NewMask(consensus.PublicKeys, consensus.PubKey)
+	if err != nil {
+		utils.GetLogInstance().Error(
```

Comment:
same as above, so many errors, please just return error.

Reply:
Same; infrequent but serious.  If our consensus key is not found in the local consensus key list, it means we are in a shard where we shouldn't be.